### PR TITLE
Feature/add metar csv2ioda

### PR DESCRIPTION
**Pull request by gthompsnJCSDA**
## Description

Addition of a new converter to handle Surface obs from METAR reports pre-formatted in comma-separated-values (CSV) into a IODA-ready netCDF4 file.  This is intended for UFO development work as a precursor to BUFR->IODA conversions of these data.  Eventually, this converter will not be needed, but it is beneficial for preliminary work to QC the incoming data from its near rawest form including variable conversions (like computing specific_humidity from dewpoint temperature and surface_pressure from altimeter setting), EMC's prepObs steps to Bounds Check and flag gross errors, etc.  The incorporated small test dataset contains only 33 distinct worldwide observations nearest 06z 2020-Oct-01.  Over 25 total years of these CSV-formatted data from mid-1996 to mid-2020 are available via NCAR's Glade-Campaign storage area.  Each hour, approximately 8000 reports worldwide are received, but the test dataset was kept small for speed/storage considerations.

### Issue(s) addressed

None

## Dependencies

None

### CODE REVIEW

File: src/metar/CMakeLists.txt
Comment (CoryMartin-NOAA) [inline src/metar/CMakeLists.txt:1]: Do we expect additional METAR converters? If not, I suggest making the directory more generic, like "conv" or "conventional" and keeping the metar_csv2ioda.py filename.

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [inline src/metar/metar_csv2ioda.py:74]: Is this a debug print or is this intended to stay permanently? Shouldn't the column names be constant?

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [inline src/metar/metar_csv2ioda.py:85]: if you can write the specific exceptions expected here, that is preferred I think.

Comment (CoryMartin-NOAA) [reply src/metar/metar_csv2ioda.py:85]: like TypeError, ValueError, etc.

Comment (CoryMartin-NOAA) [reply src/metar/metar_csv2ioda.py:85]: Also, I think you can put all of the lines that are try: except: continue into one try and except. Python will continue at the first failed try.

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [inline src/metar/metar_csv2ioda.py:120]: wdir will never be exactly zero but it will be 360? Just double checking.

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [inline src/metar/metar_csv2ioda.py:117]: Do we need this if, can't the trig function below handle this case?

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [inline src/metar/metar_csv2ioda.py:153]: Remove this test area that is commented out in a docstring.

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [inline src/metar/metar_csv2ioda.py:179]: Remove this docstring of debug lines also

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [inline src/metar/metar_csv2ioda.py:259]: Is this much verbosity necessary for non-debugging?

**Review by CoryMartin-NOAA — CHANGES_REQUESTED** [review]
Looks pretty good, but remove the debug commented out docstring areas.

File: src/metar/CMakeLists.txt
Comment (gthompsnJCSDA) [reply src/metar/CMakeLists.txt:1]: A totally fair comment.  A "conventional" subdir might be a good choice.  Other than a direct BUFR file convert of METAR to IODA via Ron's software, I also doubt there would be multiple converters.  Do you happen to know if I would need to create an entirely new github feature branch if I eliminated the metar subdir in favor of a newly named subdir, or does a folder renaming survive properly under the same branch name?  In old CVS, this would generate problems.

File: src/metar/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/metar/metar_csv2ioda.py:74]: It was effectively a debug print.  The column names only need to be constant for the variables that actually get captured (e.g., ICAO) whereas the order could be entirely different since python is capturing by name in header, not by the physical placement/order.  However, if desired, the print statement could be eliminated, although it does seem OK to leave a single print statement per running of the converter, don't you think?

File: src/metar/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/metar/metar_csv2ioda.py:85]: As a newcomer to Python programming, I don't fully follow what you're recommending. Is the solution here simply to put the consecutive "try" pieces into a single block as opposed to each one being on a separate try/except chunk?

File: src/metar/CMakeLists.txt
Comment (CoryMartin-NOAA) [reply src/metar/CMakeLists.txt:1]: As far as I know there is no problem with renaming directories. In fact, in git, if you create a blank directory, add it, but don't put any tracked files in it, it won't even track that the directory exists. So I think you're free to change the directory name in this branch/PR without issues.

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [reply src/metar/metar_csv2ioda.py:74]: Yeah it's probably fine, just checking.

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [reply src/metar/metar_csv2ioda.py:85]: Yes, if the except for each case is "continue", just do:
try:
  line1
  line2
  line3
except:
  continue

it will go to the continue the first time it encounters an error. You would only want separate try statements if the exceptions were going to be different, or if the action to perform in the except was different.

File: src/metar/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/metar/metar_csv2ioda.py:120]: This one is actually tricky and extremely particular to my own METAR processing for how I stored the data prior to reformatting to CSV file(s).  At NCAR-RAL, I created a MySQL db of METARs.  I made absolutely sure that a true north wind is stored as 360 degrees.  That was done quite purposefully to permit the wind direction to be zero degrees in the event the raw METAR reported a string like "VRB003" meaning variable wind direction at 3 knots.  This type of observation isn't DA-useful, because we cannot create unambiguous  east/northward_wind variables.  However, my reason for permitting it in MySQL and the post-created CSV file(s) is to consider a day in the distant future where a "VRB" wind direction could potentially be useful to DA.  So this is a bit of edge case just due to how I stored the data and created my own CSV files.

File: src/metar/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/metar/metar_csv2ioda.py:117]: I believe the code is needed as-is, because it protects against the possibility (though really highly unlikely) that either wind speed *or* direction is missing while the other is not missing.  It shouldn't be possible, however, the apparent double-assignment of missing ensures there's no way to have one without the other - they both get forced to missing.

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [reply src/metar/metar_csv2ioda.py:120]: Ok just checking. That is tricky.

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [reply src/metar/metar_csv2ioda.py:117]: I was referring to the if wdir==0 and wspd==0, not the is not missing

File: src/metar/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/metar/metar_csv2ioda.py:179]: A bit of conundrum in a way (to me at least).  I created all those added doc-block chunks to give 100% test coverage to the meteo_utils codes and show how they can be run using these data (METARs) as the example.  I realize it creates a large doc-block of unexecuted code, but thought it useful to others.

Would you suggest entire deletion or putting this chunk into a document or what?

File: src/metar/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/metar/metar_csv2ioda.py:259]: Agree completely. It was purely for debugging.  Similar to UFO (and all JEDI codes), is there just a ioda-converters standardized method of putting in something like oops_debug in order to get the debug prints only when a user sets an environment variable?

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [reply src/metar/metar_csv2ioda.py:179]: What about creating a separate unit test in the ioda converters just to test these (and other) utility functions?
Something that will check if, for example, a given T and P, make sure the output theta == the expected value, and so on. Can just print to a text file and check the output against a reference text file. This could be a new PR later.

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [reply src/metar/metar_csv2ioda.py:259]: I'm not sure about that, I don't know of one, but perhaps @srherbener does. I think it's unlikely because compiled vs interpreted code, so it's much easier to add/remove print statements in python scripts if you need to for quick debugging.

File: src/metar/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/metar/metar_csv2ioda.py:117]: Ah, now I got what you meant.  I coded it with efficiency in mind.  The first IF-test will rapidly get the result without the overhead of invoking the function - so this was done purposefully as a matter of speed.  I prefer it this way.

File: src/metar/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/metar/metar_csv2ioda.py:85]: OK, thanks.  I've changed my code (will push in branch once I resolve a batch of your suggestions).  And thanks for this help!!!

File: src/metar/CMakeLists.txt
Comment (gthompsnJCSDA) [reply src/metar/CMakeLists.txt:1]: Cool. Good to know.  I will go ahead with the renaming because I foresee a good likelihood of similar possible converters that I may be contributing for other conventional data prior to usage of Ron's BUFR-IODA direct.  I'm using these CSV-IODA converters to accelerate my work with UFO developments so even if these are quite temporary, I can be highly productive in preparing future UFO work/codes.

File: src/metar/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [reply src/metar/metar_csv2ioda.py:117]: I'm fine with leaving it in as is then, mainly as it's more readable/explicit, but if statements aren't free either, so the overhead will likely be minimal.

Comment (gthompsnJCSDA) [reply src/metar/metar_csv2ioda.py:117]: Agreed that an IF-block is never 'free' either, however, the block for IF will always be required anyway whereas the function call can be optional in the specific case of (0,0) values.  So there is a minute efficiency improvement as coded, not just being more explicit and readable.

File: src/metar/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/metar/metar_csv2ioda.py:259]: I think I'll keep the print line present but comment it out.  For the purposes of NOAA-GSL, I need to add a bunch more variables for their DA (elements such as cloud base height, visibility, etc.) so I would like to keep it around as I make additions and later testing/debugging.  If another reviewer comments about a standardized way to print only when in a debug-mode, then I can adjust.

**Review by CoryMartin-NOAA — APPROVED** [review]
Thanks for making the changes. I'll defer to others on the docstring issue. I think they should probably be removed now, with the intention of making a unit test for them, but I don't have a strong opposition to leaving them in now (just makes additional lines to review later when they are removed)

**Comment by gthompsnJCSDA** [comment]
> Thanks for making the changes. I'll defer to others on the docstring issue. I think they should probably be removed now, with the intention of making a unit test for them, but I don't have a strong opposition to leaving them in now (just makes additional lines to review later when they are removed)


And thank you VERY much Cory for the good review info/advice and speed of tackling.  I agree that it is worthwhile to capture in a separate test the various items in the doc-block that I was using for testing purposes.  I will go ahead and remove these chunks and do another push to branch. Thanks again!

File: src/conventional/metar_csv2ioda.py
Comment (rhoneyager) [inline src/conventional/metar_csv2ioda.py:1]: @mmiesch, @srherbener Do we need ```env python3``` or ```env python```?

File: src/conventional/metar_csv2ioda.py
Comment (rhoneyager) [inline src/conventional/metar_csv2ioda.py:23]: This file is intended to be processed through CMake. Change the extension to .py.in to reflect this.

File: src/conventional/CMakeLists.txt
Comment (rhoneyager) [inline src/conventional/CMakeLists.txt:10]: You shouldn't need to run chmod yourself. If the input file has execute permission (and this gets recorded in git), the output file should, too.

**Review by rhoneyager — CHANGES_REQUESTED** [review]
See comments re: CMake usage. Otherwise, I am fine with it.

File: src/metar/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/metar/metar_csv2ioda.py]: I went ahead and removed the doc-blocks, keeping them in a saved copy so I can refer back to them later.

File: src/conventional/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/conventional/metar_csv2ioda.py:23]: It isn't clear to me about adding the ".in" extension - there's an entire directory of 'marine' that do not have the py.in file names.  I was following others' examples.  What's the choice here - I can easily do either, but was trying to be consistent with marine subdir.

File: src/conventional/CMakeLists.txt
Comment (gthompsnJCSDA) [reply src/conventional/CMakeLists.txt:10]: I mimicked something else in the file, so I can easily remove this.

File: src/conventional/metar_csv2ioda.py
Comment (CoryMartin-NOAA) [reply src/conventional/metar_csv2ioda.py:23]: We used to end the files in .in but somewhere along the way it was changed to just be .py. See for example in the gsi-ncdiag directory. This happened in PR 263 https://github.com/JCSDA-internal/ioda-converters/pull/263 by @markjolah

File: src/conventional/metar_csv2ioda.py
Comment (srherbener) [reply src/conventional/metar_csv2ioda.py:1]: We want to explicitly declare python3 since we are moving to drop support for python2 (which was deprecated last January). We already have this captured in issue #294.

**Review by CoryMartin-NOAA — COMMENTED** [review]
I disagree that the file should be renamed to .py.in, as now it is inconsistent with the rest of the repository.

**Comment by gthompsnJCSDA** [comment]
> See comments re: CMake usage. Otherwise, I am fine with it.

I took suggestion and pushed new change.

File: src/conventional/CMakeLists.txt
Comment (gthompsnJCSDA) [reply src/conventional/CMakeLists.txt]: I eliminated the line.

**Comment by gthompsnJCSDA** [comment]
> I disagree that the file should be renamed to .py.in, as now it is inconsistent with the rest of the repository.

I personally disagree with using ".in" as the file extension.  It would be the *only* scripting language I've worked with whereby we go 'against the grain' with typical industry-practice.  However, in favor of expediency, I just made the change and re-pushed.  So if someone wants to play final arbitrator, I do have other things I can be doing instead of wasting more time.

File: src/conventional/metar_csv2ioda.py
Comment (gthompsnJCSDA) [reply src/conventional/metar_csv2ioda.py]: Fixed, so changing item to resolved.

File: src/conventional/metar_csv2ioda.py.in
Comment (srherbener) [inline src/conventional/metar_csv2ioda.py.in:14]: You should not need this now that the interpreter is explicitly declared as python3. The print syntax changed from python2 to python3 and this statement was a workaround to allow python2 print statements to use the python3 syntax.

**Comment by gthompsnJCSDA** [comment]
@rhoneyager  Is the Travis CI step failing due something I've done?  When I click Details link, it just brings me to a AWS login page asking credentials (I'm not sure I know where to find my old credentials).

File: test/CMakeLists.txt
Comment (srherbener) [inline test/CMakeLists.txt:63]: This file is missing from the ioda-converters test/testoutput directory.

**Review by srherbener — CHANGES_REQUESTED** [review]
Thanks for adding this converter. I've requested changes since there are two tests that are failing:

- `iodaconv_metar_coding_norms`
- `test_iodaconv_metar`

You can run: `ctest -VV -R iodaconv_metar_coding_norms` to get the information about what the code style checker wants changed.

The `test_iodaconv_metar` test is currently failing due to no execute permissions on the `meta_csv2ioda.py.in` script. You should be able to fix that by adding the execute permissions to the source file (`src/conventional/metar_csv2ioda.py.in`) and committing that change.

File: test/CMakeLists.txt
Comment (gthompsnJCSDA) [reply test/CMakeLists.txt:63]: Steve, is it typical to commit the output of the process to the output directory?  I was expecting that running the ctest creates the output file and therefore didn't include it specifically in the git add/commit.  I will pop it in there (in addition to your other suggestions, thanks!)

File: src/conventional/metar_csv2ioda.py.in
Comment (gthompsnJCSDA) [reply src/conventional/metar_csv2ioda.py.in:14]: OK, got it. I removed this line.

**Comment by gthompsnJCSDA** [comment]
> Thanks for adding this converter. I've requested changes since there are two tests that are failing:
> 
> * `iodaconv_metar_coding_norms`
> * `test_iodaconv_metar`
> 
> You can run: `ctest -VV -R iodaconv_metar_coding_norms` to get the information about what the code style checker wants changed.
> 
> The `test_iodaconv_metar` test is currently failing due to no execute permissions on the `meta_csv2ioda.py.in` script. You should be able to fix that by adding the execute permissions to the source file (`src/conventional/metar_csv2ioda.py.in`) and committing that change.

I am fixing this, however, I can say that I am *very* frustrated with the number of blank line complaints.  It's so different than UFO and utterly silly (TBH).  UFO code norms does not allow 2 consecutive blanks lines yet ioda expects 2 before any class declaration.  And, I have a line longer than 166 chars and I've tried google searching how to resolve it, which is easy, but what is the code-norms for line-breaking long IF-statements?  I've tried the backslash character, I've tried indenting more and more spaces as well as less and less spaces and failure every time I try something.  I'm guessing in the total dark right now how to fix it.

**Comment by gthompsnJCSDA** [comment]
I give up on the code-norm check with my long-line IF statement.  I attempted one space at a time for about 12-14 spaces and determined it's futile to guess any longer.  I found about 9 pages of stackoverflow discussing this point and it's clear that it's a moving target.  The backslash character at end of one line fails, but over and under-indenting fails and there's no clear solution (to me).

File: test/CMakeLists.txt
Comment (srherbener) [reply test/CMakeLists.txt:63]: In this repo, the test/testoutput directory is where the expected outputs go and the test/testrun directory is where the outputs from running the tests go.

**Comment by srherbener** [comment]
@gthompsnJCSDA sorry about the difficulties with the coding norms test. We are using the pycodestyle checker which is a commonly used program, but unfortunately wants a different style than the cpplint.py (from Google) checker we are using on the C++ code.

I'm taking a look at this now to see if I can get the coding norms test error fixed.

**Comment by gthompsnJCSDA** [comment]
> @gthompsnJCSDA sorry about the difficulties with the coding norms test. We are using the pycodestyle checker which is a commonly used program, but unfortunately wants a different style than the cpplint.py (from Google) checker we are using on the C++ code.
> 
> I'm taking a look at this now to see if I can get the coding norms test error fixed.


Thanks Steve!  As you can tell, it's highly frustrating to waste up to an hour or more on something so nearly pointless.  I put the "and" on either line (and coding norms complains), I changed the number of whitespaces up and down the line watching the error message say over-indent, under-indent, and neither yet still complained about the line break either before or after the "and" chunk/clause.  So it looks like a "catch-22" basically.

**Comment by srherbener** [comment]
@gthompsnJCSDA, I agree that some of the things that pycodestyle complains about seem trivial. For example I think it's perfectly clear, and okay, to put a line break after the "and" operator. I wonder what type of "gotcha" pycodestyle is trying to flag.

I was able to get the coding style errors fixed while having the metar function test pass.

I ended up removing the `metar_csv2ioda.py.in` script since CMake was still using that one. The conf_targets_deps macro will perfer the ".in" version of the file if both exist. Ie, if you a files my_script and my_script.in, conf_targets_deps will select the my_scripts.in file.

I'm sorry if I removed the wrong script, but it looked like in the discussion that everyone settled on the version without the ".in" suffix. If you need to change it to the ".in" version , use git to move the file:

`git mv metar_csv2ioda.py metar_csv2ioda.py.in`

This will avoid getting both filenames in the git repository, and preserve the history of the file.

**Comment by srherbener** [comment]
The CodeBuild clang configuration is broken right now. All the tests pass on my iMac as well as the CodeBuild gnu and intel tests. I think we can ignore the CodeBuild clang testing for now (until it gets fixed).

**Comment by srherbener** [comment]
@gthompsnJCSDA, FYI for the fixing of the coding style errors, I did the following:

The csv python package throws the exception csv.Error for any of the read errors so I specified that on the except commands. Ie, replaced `except:` with `except csv.Error:`.

I defined a data member for the reformatMetar class called float_fill and set that to the netCDF4 default fill value. Then I replaced all the calls to `netCDF4.default_fillvals['f4']` with `self.float_fill`. This shortened the long if statement down below 160 characters.

**Review by srherbener — APPROVED** [review]

**Comment by srherbener** [comment]
@rhoneyager, have your concerns been addressed now? Thanks!

**Comment by mer-a-o** [comment]
@srherbener, All tests should pass for ioda-converters. We should find a solution for the error in clang container.

**Comment by srherbener** [comment]
@mer-a-o, I agree 100%. In fact, I've submitted a ZenHub issue about this so we can track and get it fixed. I'm just saying if we want to merge this PR before the Clang CodeBuild testing is working, it should be safe to do so. Sorry about the confusion.

**Comment by gthompsnJCSDA** [comment]
> @gthompsnJCSDA, FYI for the fixing of the coding style errors, I did the following:
> 
> The csv python package throws the exception csv.Error for any of the read errors so I specified that on the except commands. Ie, replaced `except:` with `except csv.Error:`.
> 
> I defined a data member for the reformatMetar class called float_fill and set that to the netCDF4 default fill value. Then I replaced all the calls to `netCDF4.default_fillvals['f4']` with `self.float_fill`. This shortened the long if statement down below 160 characters.

Steve, I see that solution you implemented. Clearly a good temporary solution, but, honestly, I think the root cause of coding-norms problem is unchanged for a future point when a line longer than 166 chars can still occur.  It would be truly good to figure a *correct* solution down the road.

Thanks for doing your adjustments to my code to get the PR approved.  I do appreciate the direct code fixes of your own and reaching approval point for the PR. Thank yoU!

**Review by rhoneyager — APPROVED** [review]

**Comment by srherbener** [comment]
@gthompsnJCSDA I agree that the solution for the 166 character limit on line length is not a robust, systematic way to fix that issue. Other than shutting off this check, I don't really see a robust, systematic way to address going over a line length limit. I think a lot of people would agree that enforcing a maximum line length is a good thing, but I suspect getting consensus on what the maximum length should be is a lot more tenuous. I suspect that a limit of 166 characters will provide enough room for people to come up with ad hoc ways of staying within that limit.

I had another reason for doing the solution the way I did which was to make it so that if you decide in the future that you want to change the fill value, you can do it in one place instead of many places.

If you have a solution that you like better than what I did, I'm perfectly fine with you using that.

**Comment by srherbener** [comment]
The conflicts need to be resolved before merging, and if you want I'm happy to help.

**Comment by gthompsnJCSDA** [comment]
> The conflicts need to be resolved before merging, and if you want I'm happy to help.

Steve:  I thought you directly made the necessary changes to the branch?  Is that incorrect?

And, regarding line length - that is *not* what I was referencing as the real problem.  I am fully OK with a guideline for line length.  The existing problem is the proper *method* to break a long line such as mine in the potential case that a solution like yours could not fit into the line length itself.

So, let's just say an example where the IF-clauses were another 2 or 3 parenthesized AND clauses.  I have 3 'and' clauses, but if there were even more, then it would take another solution to get a line break that doesn't raise the E129 (or E127) or whatever style flag.  That is what I was truly seeking a solution.

**Comment by srherbener** [comment]
@gthompsnJCSDA, I did make the changes to this feature branch, but after those changes went in another PR (#351) got merged into develop. The develop branch has advanced and, as a result, this feature branch has conflicts with develop now.

The appearance of conflicts is something that normally happens when you have a lot of people working in the same repo. Fortunately, those people tend to be working on different parts of the repo and the emergence of conflicts tends to be occasional.

This conflict is with the `test/CMakeLists.txt` file. You can update your local clone and merge in the develop branch into your feature branch.

```
git remote update -p # update local clone

git checkout develop # update develop
git pull

git checkout feature/add_metar_csv2ioda # update feature branch
git pull

git merge develop  # merge develop branch into feature branch
```

At this point you should see a message saying that the `test/CMakeLists.txt` file has conflicts. Edit the `test/CMakeLists.txt` file. You will find marks like these that show you the conflicts.

```
<<<<< HEAD
# feature branch code from before merge
=====
# develop code from before merge
>>>>> develop
```

The HEAD section is what you had in the file before merging in develop. The develop section is what develop had before the merge. It's up to you to decide how to resolve the conflict. You can keep HEAD, or keep develop, or do some combination of code from both. 

In this case you should keep the develop code and discard the HEAD code. I took a look at the conflicts and the PR  #351 that got merged in was replacing the CMake `add_test` commands with the ecbuild `ecbuild_add_test` commands so we want to move forward with the `ecbuild_add_test` commands.

I think it will be good for you to go through this process of resolving the conflict. This won't be the last time conflicts appear so it's good to learn how to do this. I can help if you have questions.

**Comment by srherbener** [comment]
@gthompsnJCSDA, in the example you gave with the line length limit error, you could use a boolean variable to solve the problem

```
# say the if condition is broken up into 6 expressions that are all and-ed together
# the expresssion are expr1, expr2, ..., expr6
# you can use a variable to build up the expression line by line like this:
total_condition = expr1
total_condition = total_condition and expr2
total_condition = total_condition and expr3
total_condition = total_condition and expr4
total_condition = total_condition and expr5
total_condition = total_condition and expr6

if (total_condition):
    # code
```

Is this what you are looking for?

**Pull request merged**
